debugger/debugger.js

```diff
@@ -0,0 +1,310 @@
+    this.print("----------------------------------------------------------------------------------\n");
+    this.print(" ==> " + this.get_source_line(lineno - 1) + "\n");
+    this.print("----------------------------------------------------------------------------------\n");
+};
+
+Sk.Debugger.prototype.set_suspension = function(suspension) {
+    var parent = null;
+    if (!hasOwnProperty(suspension, "filename") && suspension.child instanceof Sk.misceval.Suspension) {
+        suspension = suspension.child;
+    }
+        
+    // Pop the last suspension of the stack if there is more than 0
+    if (this.suspension_stack.length > 0) {
+        this.suspension_stack.pop();
+        this.current_suspension -= 1;
+    }
+    
+    // Unroll the stack to get each suspension.
+    while (suspension instanceof Sk.misceval.Suspension) {
+        parent = suspension;
+        this.suspension_stack.push(parent);
+        this.current_suspension += 1;
+        suspension = suspension.child;
+    }
+
+    suspension = parent;
+    
+    this.print_suspension_info(suspension);
+};
+
+Sk.Debugger.prototype.add_breakpoint = function(filename, lineno, colno, temporary) {
+    var key = this.generate_breakpoint_key(filename, lineno, colno);
+    this.dbg_breakpoints[key] = new Sk.Breakpoint(filename, lineno, colno);
+    if (temporary) {
+        this.tmp_breakpoints[key] = true;
+    }
+};
+
+Sk.Debugger.prototype.suspension_handler = function(susp) {
+    return new Promise(function(resolve, reject) {
+        try {
+            resolve(susp.resume());
+        } catch(e) {
+            reject(e);
+        }
+    });
+};
+
+Sk.Debugger.prototype.resume = function() {
+    // Reset the suspension stack to the topmost
+    this.current_suspension = this.suspension_stack.length - 1;
+    
+    if (this.suspension_stack.length === 0) {
+        this.print("No running program");
+    } else {
+        var promise = this.suspension_handler(this.get_active_suspension());
+        promise.then(this.success.bind(this), this.error.bind(this));
+    }
+};
+
+Sk.Debugger.prototype.pop_suspension_stack = function() {
+    this.suspension_stack.pop();
+    this.current_suspension -= 1;
+};
+
+Sk.Debugger.prototype.success = function(r) {
+    if (r instanceof Sk.misceval.Suspension) {
+        this.set_suspension(r);
+    } else {
+        if (this.suspension_stack.length > 0) {
+            // Current suspension needs to be popped of the stack
+            this.pop_suspension_stack();
+            
+            if (this.suspension_stack.length === 0) {
+                this.print("Program execution complete");
+                return;
+            }
+            
+            var parent_suspension = this.get_active_suspension();
+            // The child has completed the execution. So override the child's resume
+            // so we can continue the execution.
+            parent_suspension.child.resume = function() {};
```

Comment:
Okay, but I'm still not convinced that this is necessary.  When foo suspends, you should get a suspension back for main, which has a child suspension for boo, which has a child suspension for goo, which has a child suspension for foo.  Then, when you want to resume, you resume the main suspension, which recursively resumes the children first, so foo will be resumed first.  If goo later suspends after foo returns because it hits another breakpoint, you should get a new suspension for main, with a child for boo, with a child for goo.

Instead, you are resuming the foo suspension manually.  Which is what causes the "problem", where the goo suspension still has a stale foo child.  I'm not sure, but I believe this can cause problems when a function actually returns a value.  I think in some circumstances, your method could lose that return value.

So, rather than keeping around the suspension stack and managing it manually, it seems like you should let the suspension infrastructure do it for you (and maybe push it onto an array each time you get a suspension if you'd prefer to use it as an array, rather than the linked list you get, when allowing the user to navigate the call stack).

This was the source of my original question.  It seems like you should not have to clear out that resume function if you use it the way it was originally intended.  So, my question was really why are you doing it this way instead of using the suspension list.  I wanted to know if I am missing something, as I still can't wrap my head around why you need to do it this way...

(Or perhaps I am totally incorrect about how the suspension infrastructure currently works...)